### PR TITLE
Graph improvements

### LINKED ISSUE #2197: Make population graphs use longs instead of ints
as the graphs have problems when a population number goes over the max int value.

### LINKED ISSUE #2883: Add a button on line graph to show it in full size
Graphs can get unreadable at certain data size. To alleviate this, a button should be added somewhere on individual graphs to bring it up to full size, preferably by using a popup.

### LINKED ISSUE #2886: Move extinct species on species list dropdown to the bottom
All extinct species in the species population graph's dropdown list should be moved to the bottom of graph for easier finding. Maybe a new extinct section can be added.

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

- Added inspection button to the graphs allowing for full size view of the graph
- Added extinction types legend to the species pop graph (only viewable in inspection popup)
- Hovering over dataset icon legend now highlights the associated dataset line
- Fixed some more bugs related to invalid minimum and maximum data value
- Fixed y value in data point tooltips having too many decimal digits
- Data points tooltip for extinct species now says the extinction type instead of 0 pop.
- Add extinct species section in the species pop graph's species list dropdown

**Related Issues**

Closes #2883 
Closes #2197 
Closes #2886

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Review by RacerBG — CHANGES_REQUESTED** [review]
1. The Physical Conditions graph's window title says "Temperature" which is incorrect - it should say "Physical Conditions".
2. A "Close" button would be nice in my opinion.
3. The new species list formatting is nice but maybe it could be improved? For example if you show explicitly which species are extinct why not show "Current Species" above the alive ones?

So far the changes are very good and they don't cause new issues at least for me.

**Comment by Kasterisk** [comment]
> The new species list formatting is nice but maybe it could be improved? For example if you show explicitly which species are extinct why not show "Current Species" above the alive ones?

The top section includes all species alive and extinct so showing "Current Species" can be misleading.

**Comment by Kasterisk** [comment]
Found out there's still some bugs left:
- Hovering over datapoints other than species pop graph's doesn't highlight the marker and show the tooltip.
- Inconsistent min datasets visible warning. It pops up in other graphs other than species pop's.

**Comment by RacerBG** [comment]
Another small thing: extinct species are shown as alive as well as extinct (due to local extinction):

**Comment by Kasterisk** [comment]
Yeah that's intended, see https://github.com/Revolutionary-Games/Thrive/issues/2886#issuecomment-983762864.

**Comment by Kasterisk** [comment]
Alright, all bugs I mentioned earlier should have been fixed now.

**Review by RacerBG — APPROVED** [review]
My concerns were adressed.

File: locale/en.po [hunk 1]
```diff
@@ -3334,108 +3347,122 @@ msgstr ""
...
-#: ../src/microbe_stage/editor/MicrobeEditorGUI.cs:1004
+#: ../src/microbe_stage/editor/MicrobeEditorGUI.cs:1026
 msgid "SPECIES_LIST"
 msgstr "Species List"
 
-#: ../src/microbe_stage/editor/MicrobeEditorGUI.cs:1040
+#: ../src/microbe_stage/editor/MicrobeEditorGUI.cs:1036
+#: ../src/microbe_stage/editor/MicrobeEditorGUI.cs:1049
+msgid "TOTAL_EXTINCTION"
+msgstr "Total Extinction"
```
Comment (hhyyrylainen) [inline locale/en.po:3400]: Shouldn't this just be Extinction? "total extinction" sounds weird to me. I think it'll be clear enough with the other translation being "local extinction".

File: scripts/update_localization.rb [hunk 2]
```diff
@@ -50,7 +50,8 @@
                   'ProcessesDescription', '-k', 'window_title', '-k', 'dialog_text', '-k',
                   'WindowTitle', '-k', 'DialogText', '-k', 'ConfirmText', '-k', 'CancelText',
                   '-k', 'ErrorMessage', '-k', 'placeholder_text', '-k', 'hint_tooltip', '-k',
-                  'TranslationServer.Translate', '-k', 'LocalizedString', '-o',
+                  'TranslationServer.Translate', '-k', 'ChartName', '-k', 'LocalizedString',
+                  '-o',
```
Comment (hhyyrylainen) [inline scripts/update_localization.rb]: This just being on its own looks pretty bad

File: locale/en.po (hunk 1 above)
Comment (RacerBG) [reply locale/en.po:3400]: Mass extinction, probably?

File: src/gui_common/charts/ChartDataSet.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        foreach (var point in dataPoints)
+        {
+            result.AddPoint((DataPoint)point.Clone());
+        }
+
+        return result;
+    }
+
+    protected void ClonePropertiesTo(ChartDataSet dataset)
```
Comment (hhyyrylainen) [inline src/gui_common/charts/ChartDataSet.cs]: Is this used for anything?

File: src/gui_common/charts/DataPoint.cs [hunk 4]
```diff
@@ -2,38 +2,37 @@
...
 ///     Note: Must be freed manually, like all Godot Node types.
 ///   </para>
 /// </remarks>
-public class DataPoint : Control
+public class DataPoint : Control, ICloneable, IEquatable<DataPoint>
 {
+#pragma warning disable SA1307 // Accessible fields should begin with upper-case letter
+    public readonly double x;
+    public readonly double y;
+#pragma warning restore SA1307 // Accessible fields should begin with upper-case letter
```
Comment (hhyyrylainen) [inline src/gui_common/charts/DataPoint.cs]: Don't we have control over naming here? Why not name these properly?

File: src/gui_common/charts/IDataSetsLegend.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+public interface IDataSetsLegend : ICloneable
+{
+    /// <summary>
+    ///   Creates datasets legend UI element. This should be called on a chart plotting/construction method.
+    /// </summary>
+    /// <param name="datasets">The datasets of a particular chart</param>
+    /// <param name="title">The legend title</param>
+    /// <returns>The legend's UI element</returns>
+    Control OnCreate(Dictionary<string, ChartDataSet> datasets, string title);
```
Comment (hhyyrylainen) [inline src/gui_common/charts/IDataSetsLegend.cs]: `CreateLegend`?

File: src/gui_common/charts/line/LineChartData.cs [hunk 6]
```diff
@@ -1,73 +1,18 @@
...
-    /// <summary>
-    ///   If this is true, visuals will be drawn (e.g lines, markers)
-    /// </summary>
-    public bool Draw
-    {
-        get => draw;
-        set
-        {
-            draw = value;
+        ClonePropertiesTo(result);
```
Comment (hhyyrylainen) [inline src/gui_common/charts/line/LineChartData.cs]: I see ClonePropertiesTo is used here, but then why is the Clone in ChartDataSet implemented using duplicate functionality (the property copying)?

File: src/microbe_stage/editor/MicrobeEditorGUI.cs [hunk 7]
```diff
@@ -903,27 +901,31 @@ public void UpdateReportTabStatistics(Patch patch)
...
             foreach (var entry in snapshot.SpeciesInPatch)
             {
-                var dataset = new LineChartData { DataColour = entry.Key.Colour };
+                var dataset = new LineChartData { Colour = entry.Key.Colour };
                 speciesPopulationChart.AddDataSet(entry.Key.FormattedName, dataset);
             }
         }
 
+        var extinctSpecies = new List<KeyValuePair<string, ChartDataSet>>();
+        var extinctPoint = new List<(string Name, DataPoint ExtinctPoint, string TimePeriod, bool ExtinctEverywhere)>();
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorGUI.cs]: `extinctPoints`?

File: src/microbe_stage/editor/MicrobeEditorGUI.cs [hunk 8]
```diff
@@ -950,6 +951,9 @@ public void UpdateReportTabStatistics(Patch patch)
                 var extinctInPatch = entry.Value <= 0;
                 var extinctEverywhere = false;
 
+                if (!anyExtinction)
+                    anyExtinction = extinctInPatch;
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorGUI.cs]:
```suggestion
                if (!extinctInPatch)
                    anyExtinction = true;
```

File: src/microbe_stage/editor/MicrobeEditorGUI.cs [hunk 9]
```diff
@@ -2425,4 +2467,49 @@ private class PendingAutoEvoPrediction
 
         public bool Finished => AutoEvoRun.Finished;
     }
+
+    /// <summary>
+    ///   Extension of LineChart's default datasets dropdown legend to allow sectioning of extinct species.
+    /// </summary>
+    private class SpeciesPopDatasetsLegend : LineChart.DataSetsDropdownLegend
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorGUI.cs]: I'd put this in a separate file to try to keep down the length of this file, as this has now reached 2500 lines.

File: locale/en.po (hunk 1 above)
Comment (hhyyrylainen) [reply locale/en.po:3400]: Well that's what comes to mind. But I'm pretty sure this is related to a single species going extinct from the planet (which is how it is worded in the auto-evo report). I think the graphs should use similar terminology, if the short "Extinction" text is not acceptable.

Comment (RacerBG) [reply locale/en.po:3400]: Both terms are interchangeable in my opinion. Either total or mass extinction could be used for a single species or a couple of species but as you said something like "extinct from patch" and "extinct from planet" similar to Auto-Evo would be better.

Comment (Kasterisk) [reply locale/en.po]: Changed this to "Extinct from the planet" and "Extinct from patch".

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I think the code is now sufficiently cleaned up to be accepted, at least my brain can't right now take another full look at it.
I did some playtesting and it's mostly good. I noticed just two a bit weird things.
First is that a species is in the extinct list even though it isn't currently extinct:





And I'd like to see a small tweak to the "close" button layout. I think it would make sense to place it on the same line as the legend explanations, just right aligned. That way there's slightly more vertical space for the graph to be bigger.

**Comment by Kasterisk** [comment]
Should be fixed.

**Comment by Kasterisk** [comment]
I tinkered around with the lines in hope of fixing the ugly aliasing and I finally got it to look smoother by following Calinou's workaround in this [reddit post](https://www.reddit.com/r/godot/comments/k7qcwa/comment/geukkqe/?utm_source=share&utm_medium=web2x&context=3).





Fun fact, the lines were actually translucent before due to directly taking the cell species color which has some transparency in them. This is probably one of the reason why the lines look off the last time I tried enabling antialiasing.

Also a bonus, the lines and markes are now made slightly larger in the expanded view.

**Review by hhyyrylainen — APPROVED** [review]
I think my concerns were addressed, and gameplay testing didn't turn up any further issues (and the ones I brought up before are fixed).

**Pull request merged**